### PR TITLE
CADC-8776 check the doiadmin cert owns the DOI parent node

### PR DESCRIPTION
**Pull request by jburke-cadc**

### CODE REVIEW

File: doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java [hunk 1]
```diff
@@ -251,12 +264,76 @@ private static MultiValuedProperties getConfig(boolean verify) {
...
     }
 
+    // check that the DOI parent node path, configured with the V0SPACE_PARENT_PATH_KEY property,
+    // exists and has the expected properties.
+    private static void checkParentFolders() {
+        MultiValuedProperties config = getConfig();
+        URI vospaceResourceID = URI.create(config.getFirstPropertyValue(DoiInitAction.VAULT_RESOURCE_ID_KEY));
+        String parentPath = config.getFirstPropertyValue(PARENT_PATH_KEY);
+
+        Subject adminSubject = SSLUtil.createSubject(new File("/config/doiadmin.pem"));
```
Comment (andamian) [inline doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java:281]: Maybe the `README` should mention the exact location where the certs are expected?

File: doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java [hunk 2]
```diff
@@ -251,12 +264,76 @@ private static MultiValuedProperties getConfig(boolean verify) {
...
+        adminSubject = AuthenticationUtil.augmentSubject(adminSubject);
+        String adminUsername = getUsername(adminSubject);
+
+        VOSpaceClient vosClient = new VOSpaceClient(vospaceResourceID);
+        String currentPath = "";
+        String[] paths = parentPath.split("/");
+        for (String path : paths) {
```
Comment (andamian) [inline doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java:288]: Is this recursive checking required? It is performed on the server side as well and reported back to the client.

File: doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java [hunk 3]
```diff
@@ -251,12 +264,76 @@ private static MultiValuedProperties getConfig(boolean verify) {
...
+            }
+
+            // check inheritPermissions is true (does inheritPermissions need to be true?)
+            if (!containerNode.inheritPermissions) {
+                throw new IllegalStateException(String.format("node %s must have inheritPermissions set to true", path));
+            }
+        }
+    }
+
+    private static String getUsername(Subject subject) {
```
Comment (andamian) [inline doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java:328]: Hmm. This is almost like `AuthenticationUtil.getUseridsFromSubject()` but not quite...

File: doi/src/main/java/ca/nrc/cadc/doi/GetAction.java [hunk 4]
```diff
@@ -127,7 +126,7 @@ private Title getTitle(Resource resource) {
         Title title = null;
         List<Title> titles = resource.getTitles();
         for (Title t : titles) {
-            if (t.titleType == null) {
+            if (StringUtil.hasText(t.getValue())) {
```
Comment (andamian) [inline doi/src/main/java/ca/nrc/cadc/doi/GetAction.java:129]: So `titleType` is a but.

File: doi/src/main/java/ca/nrc/cadc/doi/PostAction.java [hunk 5]
```diff
@@ -270,20 +272,20 @@ private void makeDOIFindable(ContainerNode doiContainerNode)
...
         log.debug("content: " + content);
         InputStream inputStream = new ByteArrayInputStream(content.getBytes());
 
         // upload
         HttpUpload put = new HttpUpload(inputStream, doiURL);
         put.setRequestProperty("Authorization", "Basic " + Base64.encodeString(getCredentials()));
         put.setBufferSize(64 * 1024);
-        put.setContentType("text/plain;charset=UTF-8");
-        put.run();
+        put.setRequestProperty("Content-Type", "text/plain;charset=UTF-8");
```
Comment (andamian) [inline doi/src/main/java/ca/nrc/cadc/doi/PostAction.java:283]: `HttpTransfer.CONTENT_TYPE`?

File: doi/src/main/java/ca/nrc/cadc/doi/ServiceAvailability.java [hunk 6]
```diff
@@ -156,9 +165,13 @@ public Availability getStatus() {
                 log.debug("not configured: " + Standards.UMS_USERS_10);
             }
 
-            URI groupsURI = null;
+            URI groupsURI;
             try {
-                groupsURI = localAuthority.getServiceURI(Standards.GMS_SEARCH_10.toString());
+                Set<URI> groupsURIs = localAuthority.getServiceURIs(Standards.GMS_SEARCH_10);
```
Comment (andamian) [inline doi/src/main/java/ca/nrc/cadc/doi/ServiceAvailability.java:170]: Not sure why you need the GMS configured in the `doi.properties` when you use the one in `cadc-registry.properties`. Am I missing something?

File: doi/src/intTest/java/ca/nrc/cadc/doi/IntTestBase.java [hunk 7]
```diff
@@ -177,9 +177,14 @@ protected String createDOI(Subject testSubject)
...
         Map<String, Object> params = new HashMap<>();
-        FileContent fileContent = new FileContent(doiXML, "text/xml", StandardCharsets.UTF_8);
-        params.put("doiMetadata", fileContent);
-        params.put("journalref", journalRef == null ? "" : journalRef);
+        if (StringUtil.hasText(doiXML)) {
+            FileContent fileContent = new FileContent(doiXML, "text/xml", StandardCharsets.UTF_8);
+            params.put("doiMetadata", fileContent);
+        }
+        if (StringUtil.hasText(journalRef)) {
+            params.put("journalref", journalRef == null ? "" : journalRef);
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply doi/src/intTest/java/ca/nrc/cadc/doi/IntTestBase.java]: There is one test where the journalRef is null, but the IF condition negates journalRef from being null in the IF. Still that check and set of the journalRef isn't right. It's okay for the current int tests, but you should be able to post a journalRef with an empty value, which translates in the PostAction to deleting the existing journalRef. You can't post an empty journalReg with that test method. Checking that the journalRef is not null before adding it as a parameter is all that's needed. Fixed.

File: doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java (hunk 1 above)
Comment (jburke-cadc) [reply doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java:281]: That would be helpful, fixed.

File: doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java (hunk 2 above)
Comment (jburke-cadc) [reply doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java:290]: That's a good question. We use two container nodes as the root for the DOI's: **/AstroDataCitationDOI/CISTI.CANFAR**, but the owner and permissions only matter on the second node. I do remember something about why two parent nodes, and why owner and permissions matter for both, but I can't find any reference. Tempted to change that check for only the second parent node, but I'll leave it until I'm more sure. Added a TODO there so I don't forget.

File: doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java (hunk 3 above)
Comment (jburke-cadc) [reply doi/src/main/java/ca/nrc/cadc/doi/DoiInitAction.java:330]: Yeah close, but just not quite.

File: doi/src/main/java/ca/nrc/cadc/doi/GetAction.java (hunk 4 above)
Comment (jburke-cadc) [reply doi/src/main/java/ca/nrc/cadc/doi/GetAction.java:129]: titleType is optional, but the title value went from empty not allowed in the old schema, to empty allowed in the new schema. We don't want empty titles, hence the check. Don't know why it was checking for an optional attribute previously, doesn't make sense.

File: doi/src/main/java/ca/nrc/cadc/doi/PostAction.java (hunk 5 above)
Comment (jburke-cadc) [reply doi/src/main/java/ca/nrc/cadc/doi/PostAction.java:283]: Works for me.

File: doi/src/main/java/ca/nrc/cadc/doi/ServiceAvailability.java (hunk 6 above)
Comment (jburke-cadc) [reply doi/src/main/java/ca/nrc/cadc/doi/ServiceAvailability.java:170]: Removed the GMS configuration, no longer needed. Removed the int test properties file, and added a intTest README to explain the int test setup.

**Review by andamian — APPROVED** [review]

**Pull request merged**